worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java

```diff
@@ -197,28 +201,61 @@ private void handleTaskDataInjection(final TaskMessage taskMessage, final long i
+            try {
+                offloadedTaskData = retrieveTaskDataFromStore(taskMessageStorageRefOpt.get(), inboundMessageId);
+            } catch (final ReferenceNotFoundException e) {
+                throw new InvalidDeliveryException(e.getMessage(), inboundMessageId);
+            }
             taskMessage.setTaskData(offloadedTaskData);
         }
         // If hasTaskData and !hasStorageRef, nothing to do
     }
 
     private byte[] retrieveTaskDataFromStore(final String taskMessageStorageRef, final long inboundMessageId) 
-            throws InvalidDeliveryException, TransientDeliveryException
+            throws ReferenceNotFoundException, TransientDeliveryException
     {
         try (final var inputStream = dataStore.retrieve(taskMessageStorageRef)) {
             final var taskData = inputStream.readAllBytes();
             offloadedPayloadsToDelete.put(inboundMessageId, taskMessageStorageRef);
             return taskData;
-        } catch (final ReferenceNotFoundException ex) {
-            throw new InvalidDeliveryException("TaskMessage's TaskData could not be retrieved from DataStore", 
-                    inboundMessageId, ex);
         } catch (final IOException | DataStoreException ex) {
+            if (ex instanceof ReferenceNotFoundException) {
+                throw (ReferenceNotFoundException)ex;
+            }
             throw new TransientDeliveryException(
                 "TaskMessage's TaskData could not be retrieved from DataStore", inboundMessageId, ex);
         }
     }
 
+    private void handleInvalidDelivery(
+        final long inboundMessageId,
+        final TaskMessage taskMessage,
+        final byte[] deliveryMessageData,
+        final Map<String, Object> deliveryHeaders,
+        final String invalidDeliveryExceptionMessage
+    )
+    {
+        try {
+            final RabbitTaskInformation taskInformation = new RabbitTaskInformation(String.valueOf(inboundMessageId), true);
+            taskInformation.incrementResponseCount(true);
+            final var publishHeaders = new HashMap<>(deliveryHeaders);
+            publishHeaders.put(RABBIT_HEADER_CAF_WORKER_INVALID, invalidDeliveryExceptionMessage);
+
+            publisherEventQueue.add(new WorkerPublishQueueEvent(deliveryMessageData, invalidRoutingKey, taskInformation, publishHeaders));
+
+            final var trackingMessage = trackingMessageCreator.createInvalidTaskMessage(
+                taskMessage, invalidDeliveryExceptionMessage, invalidRoutingKey, workerConfiguration);
+            final var serializedTrackingMessage = codec.serialise(trackingMessage);
+            publisherEventQueue.add(new WorkerPublishQueueEvent(
+                serializedTrackingMessage, invalidRoutingKey, taskInformation, publishHeaders)
+            );
+        } catch (CodecException e) {
+            LOG.error("Failed to serialise report update task data.");
+            throw new RuntimeException(e);
+        }
+    }
```

Comment:
Both messages to the publisherEventQueue, and the trackingMessage have the invalidRoutingKey as the queue, this is as per the WorkerTaskImpl.  Is that correct?